File: examples/cpp/range_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct square_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename tsl::reg_param<Vec>::type value) const {
        return tsl::mul<Vec>(value, value);
    }
};

struct add_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::add<Vec>(left, right);
    }
};

struct less_than_op {
    template <class Vec>
    typename Vec::mask_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::less_than<Vec>(left, right);
    }
};

struct negative_op {
    template <class Vec>
    typename Vec::mask_type operator()(typename tsl::reg_param<Vec>::type value) const {
        return tsl::less_than<Vec>(
            value,
            tsl::set1<Vec>(static_cast<typename Vec::base_type>(0)));
    }
};

struct masked_pair_sum_op {
    std::int64_t total = 0;

    template <class Vec>
    void operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) {
        const auto zero = tsl::set1<Vec>(static_cast<typename Vec::base_type>(0));
        const auto sum = tsl::add<Vec>(left, right);
        const auto selected = tsl::blend<Vec>(active, zero, sum);
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(selected));
    }

    std::int64_t finalize() const {
        return total;
    }
};

struct masked_sum_sink {
    std::int64_t total = 0;

    template <class Vec>
    void operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type value) {
        const auto zero = tsl::set1<Vec>(static_cast<typename Vec::base_type>(0));
        const auto selected = tsl::blend<Vec>(active, zero, value);
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(selected));
    }
};

struct chunk_sum_op {
    const std::int32_t* base = nullptr;
    std::int64_t total = 0;
    std::size_t visited = 0;
    bool metadata_ok = true;

    template <class Vec>
    void operator()(
        const std::int32_t* ptr,
        std::size_t offset,
        std::size_t count) {
        if (ptr != base + offset) {
            metadata_ok = false;
        }
        const auto values = tsl::load<Vec, false>(ptr);
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(values));
        visited += count;
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>((i * 31) % 79) - 39);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 17) % 61) - 30);
    }
}

std::int64_t sum_values(const std::vector<std::int32_t>& values) {
    std::int64_t total = 0;
    for (const auto value : values) {
        total += value;
    }
    return total;
}

std::int64_t expected_masked_pair_sum(
    const std::vector<std::int32_t>& left,
    const std::vector<std::int32_t>& right) {
    std::int64_t total = 0;
    for (std::size_t i = 0; i < left.size(); ++i) {
        if (left[i] < right[i]) {
            total += static_cast<std::int64_t>(left[i]) + right[i];
        }
    }
    return total;
}

bool verify_transform_unary(
    const std::vector<std::int32_t>& input,
    const std::vector<std::int32_t>& output) {
    for (std::size_t i = 0; i < input.size(); ++i) {
        if (output[i] != input[i] * input[i]) {
            return false;
        }
    }
    return true;
}

bool verify_transform_binary(
    const std::vector<std::int32_t>& left,
    const std::vector<std::int32_t>& right,
    const std::vector<std::int32_t>& output) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        if (output[i] != left[i] + right[i]) {
            return false;
        }
    }
    return true;
}

bool run_range_case() {
    constexpr std::size_t count = 1003;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    std::vector<std::int32_t> output(count);
    std::vector<std::int32_t> selected(count, 1234567);
    fill_inputs(left, right);

    tsl::algo::transform_unary<4, tsl::algo::alignment::unaligned>(
        square_op{}, left, output);
    if (!verify_transform_unary(left, output)) {
        return false;
    }

    tsl::algo::transform_binary<4, tsl::algo::alignment::unaligned>(
        add_op{}, left, right, output);
    if (!verify_transform_binary(left, right, output)) {
        return false;
    }

    using mask_type = tsl::algo::fixed_integral_mask_type<4, std::int32_t>;
    std::vector<mask_type> masks(
        tsl::algo::integral_mask_chunk_count<4, std::int32_t>(count));
    const auto mask_chunks = tsl::algo::predicate_binary<
        4,
        tsl::algo::alignment::unaligned>(
        less_than_op{},
        left,
        right,
        masks);
    if (mask_chunks != masks.size()) {
        return false;
    }

    const auto produced = tsl::algo::select_masked_unary<
        4,
        tsl::algo::alignment::unaligned>(
        negative_op{},
        left,
        masks,
        selected);
    std::size_t expected_selected = 0;
    for (std::size_t i = 0; i < count; ++i) {
        if ((left[i] < right[i]) && (left[i] < 0)) {
            if (selected[expected_selected] != left[i]) {
                return false;
            }
            expected_selected += 1;
        }
    }
    if (produced != expected_selected) {
        return false;
    }

    const auto aggregate = tsl::algo::aggregate_masked_binary<
        4,
        tsl::algo::alignment::unaligned>(
        masked_pair_sum_op{},
        left,
        right,
        masks);
    if (aggregate != expected_masked_pair_sum(left, right)) {
        return false;
    }

    masked_sum_sink sink;
    tsl::algo::consume_masked_unary<4, tsl::algo::alignment::unaligned>(
        sink,
        left,
        masks);

    std::int64_t expected_sink = 0;
    for (std::size_t i = 0; i < count; ++i) {
        if (left[i] < right[i]) {
            expected_sink += left[i];
        }
    }
    if (sink.total != expected_sink) {
        return false;
    }

    chunk_sum_op chunk_sum{left.data()};
    tsl::algo::for_each_chunk<4>(chunk_sum, left);
    return chunk_sum.metadata_ok &&
           (chunk_sum.visited == left.size()) &&
           (chunk_sum.total == sum_values(left));
}

int main() {
    return run_range_case() ? 0 : 1;
}

File: examples/rust/src/bin/range_operator.rs
use tsl::profile;
use tsl::tsl_core::StaticSimdVector;

struct Square;

impl<V> profile::algo::UnaryKernel<V> for Square
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::MulImpl,
    V::RegisterType: Copy,
{
    fn apply(&mut self, value: V::RegisterType) -> V::RegisterType {
        profile::mul::<V>(value, value)
    }
}

struct Add;

impl<V> profile::algo::BinaryKernel<V> for Add
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::AddImpl,
{
    fn apply(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::RegisterType {
        profile::add::<V>(left, right)
    }
}

struct LessThan;

impl<V> profile::algo::BinaryPredicateKernel<V> for LessThan
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::Less_thanImpl,
{
    fn test(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::MaskType {
        profile::less_than::<V>(left, right)
    }
}

struct Negative;

impl<V> profile::algo::UnaryPredicateKernel<V> for Negative
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::Less_thanImpl
        + profile::detail::primitives::Set1Impl,
{
    fn test(&mut self, value: V::RegisterType) -> V::MaskType {
        profile::less_than::<V>(value, profile::set1::<V>(0))
    }
}

struct MaskedPairSum {
    total: i64,
}

impl<V> profile::algo::MaskedBinaryAggregateKernel<V> for MaskedPairSum
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::AddImpl
        + profile::detail::primitives::SelectImpl
        + profile::detail::primitives::HaddImpl
        + profile::detail::primitives::Set1Impl,
{
    type Output = i64;

    fn accumulate(&mut self, active: V::MaskType, left: V::RegisterType, right: V::RegisterType) {
        let zero = profile::set1::<V>(0);
        let sum = profile::add::<V>(left, right);
        let selected = profile::select::<V>(active, sum, zero);
        self.total += i64::from(profile::hadd::<V>(selected));
    }

    fn finalize(&self) -> Self::Output {
        self.total
    }
}

struct MaskedSumSink {
    total: i64,
}

impl<V> profile::algo::MaskedUnaryConsumeKernel<V> for MaskedSumSink
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::SelectImpl
        + profile::detail::primitives::HaddImpl
        + profile::detail::primitives::Set1Impl,
{
    fn consume(&mut self, active: V::MaskType, value: V::RegisterType) {
        let zero = profile::set1::<V>(0);
        let selected = profile::select::<V>(active, value, zero);
        self.total += i64::from(profile::hadd::<V>(selected));
    }
}

struct ChunkSum {
    base: *const i32,
    total: i64,
    visited: usize,
    metadata_ok: bool,
}

impl ChunkSum {
    fn new(input: &[i32]) -> Self {
        Self {
            base: input.as_ptr(),
            total: 0,
            visited: 0,
            metadata_ok: true,
        }
    }
}

impl<V> profile::algo::ChunkKernel<V> for ChunkSum
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::LoadImpl<false>
        + profile::detail::primitives::HaddImpl,
{
    unsafe fn apply(&mut self, ptr: *const V::BaseType, offset: usize, count: usize) {
        let expected_ptr = unsafe { self.base.add(offset) };
        if ptr != expected_ptr || count != V::ELEMENT_COUNT {
            self.metadata_ok = false;
        }

        let values = unsafe { profile::load::<V, false>(ptr) };
        self.total += i64::from(profile::hadd::<V>(values));
        self.visited += count;
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = ((i * 31) % 79) as i32 - 39;
        *right_value = ((i * 17) % 61) as i32 - 30;
    }
}

fn sum_values(values: &[i32]) -> i64 {
    values.iter().map(|value| i64::from(*value)).sum()
}

fn expected_masked_pair_sum(left: &[i32], right: &[i32]) -> i64 {
    left.iter()
        .zip(right)
        .filter_map(|(&left_value, &right_value)| {
            (left_value < right_value).then_some(i64::from(left_value) + i64::from(right_value))
        })
        .sum()
}

fn main() {
    let count = 1003usize;
    let mut left = vec![0i32; count];
    let mut right = vec![0i32; count];
    let mut output = vec![0i32; count];
    let mut selected = vec![i32::MAX; count];
    fill_inputs(&mut left, &mut right);

    let policy = tsl::dataparallel::generic::<4>();

    let mut square = Square;
    profile::algo::transform_unary(policy, &mut square, &left, &mut output);
    for (actual, input) in output.iter().zip(left.iter()) {
        assert_eq!(*actual, *input * *input);
    }

    let mut add = Add;
    profile::algo::transform_binary(policy, &mut add, &left, &right, &mut output);
    for ((actual, left_value), right_value) in output.iter().zip(left.iter()).zip(right.iter()) {
        assert_eq!(*actual, *left_value + *right_value);
    }

    let mask_count = profile::algo::integral_mask_chunk_count::<_, i32>(policy, count);
    let mut masks = vec![0u64; mask_count];
    let mut less_than = LessThan;
    let produced_masks =
        profile::algo::predicate_binary(policy, &mut less_than, &left, &right, &mut masks);
    assert_eq!(produced_masks, masks.len());

    let mut negative = Negative;
    let produced =
        profile::algo::select_masked_unary(policy, &mut negative, &left, &masks, &mut selected);
    let mut expected_selected = 0usize;
    for i in 0..count {
        if left[i] < right[i] && left[i] < 0 {
            assert_eq!(selected[expected_selected], left[i]);
            expected_selected += 1;
        }
    }
    assert_eq!(produced, expected_selected);
    for value in selected.iter().skip(produced) {
        assert_eq!(*value, i32::MAX);
    }

    let mut aggregate = MaskedPairSum { total: 0 };
    let aggregate_result =
        profile::algo::aggregate_masked_binary(policy, &mut aggregate, &left, &right, &masks);
    assert_eq!(aggregate_result, expected_masked_pair_sum(&left, &right));

    let mut sink = MaskedSumSink { total: 0 };
    profile::algo::consume_masked_unary(policy, &mut sink, &left, &masks);
    let expected_sink: i64 = left
        .iter()
        .zip(right.iter())
        .filter_map(|(&left_value, &right_value)| {
            (left_value < right_value).then_some(i64::from(left_value))
        })
        .sum();
    assert_eq!(sink.total, expected_sink);

    let mut chunk_sum = ChunkSum::new(&left);
    profile::algo::for_each_chunk(policy, &mut chunk_sum, &left);
    assert!(chunk_sum.metadata_ok);
    assert_eq!(chunk_sum.visited, left.len());
    assert_eq!(chunk_sum.total, sum_values(&left));
}
